cpp/src_prims/metrics/trustworthiness_score.cuh

```diff
@@ -0,0 +1,193 @@
+void run_knn(const raft::handle_t &h, math_t *input, int n, int d,
+             int n_neighbors, int64_t *indices, math_t *distances) {
+  std::vector<float *> ptrs(1);
+  std::vector<int> sizes(1);
+  ptrs[0] = input;
+  sizes[0] = n;
+
+  raft::spatial::knn::brute_force_knn(h, ptrs, sizes, d, input, n, indices,
+                                      distances, n_neighbors, true, true,
+                                      nullptr, distance_type);
+}
+
+/**
+     * @brief Compute the trustworthiness score
+     * @param h Raft handle
+     * @param X: Data in original dimension
+     * @param X_embedded: Data in target dimension (embedding)
+     * @param n: Number of samples
+     * @param m: Number of features in high/original dimension
+     * @param d: Number of features in low/embedded dimension
+     * @param n_neighbors Number of neighbors considered by trustworthiness score
+     * @param batchSize Batch size
+     * @return Trustworthiness score
+     */
+template <typename math_t, raft::distance::DistanceType distance_type>
+double trustworthiness_score(const raft::handle_t &h, math_t *X,
+                             math_t *X_embedded, int n, int m, int d,
+                             int n_neighbors, int batchSize = 512) {
+  cudaStream_t stream = h.get_stream();
+
+  const int KNN_ALLOC = n * (n_neighbors + 1);
+  rmm::device_uvector<int64_t> emb_ind(KNN_ALLOC, stream);
+  rmm::device_uvector<math_t> emb_dist(KNN_ALLOC, stream);
+
+  run_knn<distance_type>(h, X_embedded, n, d, n_neighbors + 1, emb_ind.data(),
+                         emb_dist.data());
+
+  const int PAIRWISE_ALLOC = batchSize * n;
+  rmm::device_uvector<int> X_ind(PAIRWISE_ALLOC, stream);
+  rmm::device_uvector<math_t> X_dist(PAIRWISE_ALLOC, stream);
+  rmm::device_uvector<math_t> X_sorted_dist(PAIRWISE_ALLOC, stream);
+
+  double t = 0.0;
+  rmm::device_uvector<double> t_dbuf(1, stream);
+  double *d_t = t_dbuf.data();
+
+  int toDo = n;
+  while (toDo > 0) {
+    int curBatchSize = min(toDo, batchSize);
+
+    // Takes at most batchSize vectors at a time
+
+    size_t workspaceSize = 0;
+
+    raft::distance::distance<distance_type, math_t, math_t, math_t>(
+      &X[(n - toDo) * m], X, X_dist.data(), curBatchSize, n, m, (void *)nullptr,
+      workspaceSize, stream);
+    CUDA_CHECK(cudaPeekAtLastError());
+
+    size_t colSortWorkspaceSize = 0;
+    bool bAllocWorkspace = false;
+
+    MLCommon::Selection::sortColumnsPerRow(
+      X_dist.data(), X_ind.data(), curBatchSize, n, bAllocWorkspace, nullptr,
+      colSortWorkspaceSize, stream, X_sorted_dist.data());
+
+    if (bAllocWorkspace) {
+      rmm::device_uvector<char> sortColsWorkspace(colSortWorkspaceSize, stream);
+
+      MLCommon::Selection::sortColumnsPerRow(
+        X_dist.data(), X_ind.data(), curBatchSize, n, bAllocWorkspace,
+        sortColsWorkspace.data(), colSortWorkspaceSize, stream,
+        X_sorted_dist.data());
+    }
+    CUDA_CHECK(cudaPeekAtLastError());
```

Comment:
@cjnolet this portion of the code is unclear to me. Is it fine or should it be updated?

Reply:
At a glance it looks right to me. `sortColumnsPerRow` will perform the computation in the first call if the dimensions are small enough, otherwise it will compute the size of the workspace to be allocated and will require an additional call to perform the computation.